lib/sms/sms_submit.c

```diff
@@ -0,0 +1,360 @@
+	int err = 0;
+	static uint8_t concat_msg_id = 1;
+	static uint8_t message_ref = 1;
+
+	uint8_t size = 0;
+	uint16_t text_size = strlen(text);
+	uint8_t encoded[SMS_MAX_PAYLOAD_LEN_CHARS];
+	uint8_t encoded_data_size_octets = 0;
+	uint8_t encoded_data_size_septets = 0;
+	const uint8_t udh[] = {0x05, 0x00, 0x03, 0x01, 0x01, 0x01, 0x00};
+	char ud[SMS_MAX_PAYLOAD_LEN_CHARS];
+
+	memset(encoded, 0, SMS_MAX_PAYLOAD_LEN_CHARS);
+	memcpy(ud, udh, sizeof(udh));
+
+	uint16_t text_encoded_size = 0;
+	uint8_t concat_seq_number = 0;
+	char *text_index = text;
+	char *send_bufs[CONFIG_SMS_SEND_CONCATENATED_MSG_MAX_CNT] = {0};
+	uint16_t send_bufs_ud_pos[CONFIG_SMS_SEND_CONCATENATED_MSG_MAX_CNT] = {0};
+
+	while (text_encoded_size < text_size) {
+		if (concat_seq_number >= CONFIG_SMS_SEND_CONCATENATED_MSG_MAX_CNT) {
+			LOG_WRN("Sent data cannot fit into maximum number of concatenated messages (%d)",
+				CONFIG_SMS_SEND_CONCATENATED_MSG_MAX_CNT);
+			err = -E2BIG;
+			goto error;
+		}
+
+		send_bufs[concat_seq_number] = k_malloc(SMS_AT_CMGS_BUF_SIZE);
+		if (send_bufs[concat_seq_number] == NULL) {
+			LOG_ERR("Unable to send concatenated message due to no memory");
+			err = -ENOMEM;
+			goto error;
+		}
+
+		uint16_t text_part_size = MIN(strlen(text_index), 153);
+
+		memcpy(ud + SMS_UDH_CONCAT_SIZE_SEPTETS, text_index, text_part_size);
+
+		size = string_conversion_ascii_to_gsm7bit(ud, sizeof(udh) + text_part_size,
+			encoded, &encoded_data_size_octets, &encoded_data_size_septets, true);
+
+		text_encoded_size += size - sizeof(udh);
+		text_index += size - sizeof(udh);
+
+		sms_submit_encode(
+			send_bufs[concat_seq_number],
+			encoded_number,
+			encoded_number_size,
+			encoded_number_size_octets,
+			encoded + SMS_UDH_CONCAT_SIZE_OCTETS,
+			encoded_data_size_octets - SMS_UDH_CONCAT_SIZE_OCTETS,
+			encoded_data_size_septets,
+			message_ref,
+			SMS_UDH_CONCAT_SIZE_OCTETS,
+			&send_bufs_ud_pos[concat_seq_number]);
+
+		message_ref++;
+		concat_seq_number++;
+	}
+
+	for (int i = 0; i < concat_seq_number; i++) {
+		char udh_str[13] = {0};
+		enum at_cmd_state state = 0;
+
+		sprintf(udh_str, "050003%02X%02X%02X", concat_msg_id, concat_seq_number, i + 1);
+
+		memcpy(send_bufs[i] + send_bufs_ud_pos[i], udh_str, strlen(udh_str));
+
+		err = at_cmd_write(send_bufs[i], at_response_str, sizeof(at_response_str), &state);
+		if (err) {
+			LOG_ERR("at_cmd_write returned state=%d, err=%d", state, err);
+			goto error;
+		}
+		LOG_DBG("AT Response:%s", log_strdup(at_response_str));
+
+		/* Just looping without threading seems to work fine and we don't need to wait
+		 * for CDS response. Otherwise we would need to send 2nd message from work queue
+		 * and store a lot of state information.
+		 */
+	}
+
+error:
+	for (int i = 0; i < concat_seq_number; i++) {
+		k_free(send_bufs[i]);
+	}
+	concat_msg_id++;
+	return err;
+}
+
+int sms_submit_send(char *number, char *text)
+{
+	char at_response_str[SMS_AT_RESPONSE_MAX_LEN];
```

Comment:
Personally I do recommend to put variables at the beginning of a function to have a better overview of the stack and to avoid mixing code and declarations. This function puts the following on the stack:
```
char at_response_str[SMS_AT_RESPONSE_MAX_LEN];
char empty_string[] = "";
int ret;
enum at_cmd_state state = 0;
uint8_t encoded_number[SMS_MAX_ADDRESS_LEN_CHARS + 1];
uint8_t encoded_number_size = strlen(number);
uint8_t encoded_number_size_octets = SMS_MAX_ADDRESS_LEN_CHARS + 1;
uint8_t size = 0;
uint16_t text_size = strlen(text);
uint8_t encoded[SMS_MAX_PAYLOAD_LEN_CHARS];
uint8_t encoded_data_size_octets = 0;
uint8_t encoded_data_size_septets = 0;
char send_buf[SMS_AT_CMGS_BUF_SIZE];
```
I have not made an exact count, but already:
```
char at_response_str[SMS_AT_RESPONSE_MAX_LEN]; // 256
char send_buf[SMS_AT_CMGS_BUF_SIZE]; // 400
uint8_t encoded[SMS_MAX_PAYLOAD_LEN_CHARS]; // 160
```
are 816 bytes of stack; that's a lot. This is the case for several functions in this library. If a function uses that much stack, it needs to be documented at least, but I would rather try to reduce the stack size by declaring some of those buffers as static, like suggested elsewhere.

Reply:
Let's take the general discussion under this item. So yes, we can certainly put many of the variables as static if that's preferred. Do we have any documentation on guidance related to flash vs stack usage?

Reply:
The reason I've used stack instead of static variables were memory issues we had with modem shell. Those have been resolved now. I guess the question is whether we require more stack vs RAM as RAM usage will increase significantly if I add many of the bigger variables as static.

Reply:
And related to the location of the declarations. Generally, I've been taught to put the declaration before the variable is needed to limit the scope and I've been using that practice although it may not be the most preferred thing in C, especially from the stack usage perspective.

Reply:
Variables moved at least mainly to the beginning of the functions.
Stack usage reduce significantly by using static temporary buffer (sms_buf_tmp) which is used in various places to save stack.
Overall, SRAM use has now increased 1088 bytes and that's all we need. We can do even the heap optimization for concatenated message sending with this memory use.

Reply:
This should be fine now. There is just one 160 byte buffer in the stack for sending (user_data variable). The rest are in static buffers that are reused in receiving side also to save memory.